src/Features/CSharp/Portable/CodeRefactorings/DeclareAsNullable/DeclareAsNullableCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Text;
+using System.Diagnostics;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.DeclareAsNullable
+{
+    /// <summary>
+    /// If you apply a null test on a symbol that isn't nullable, then we'll help you make that symbol nullable.
+    /// For example: `nonNull == null`, `nonNull?.Property`
+    /// </summary>
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.DeclareAsNullable), Shared]
+    internal class DeclareAsNullableCodeRefactoringProvider : CodeRefactoringProvider
```

Comment:
Why a refactoring? I think fix-all is most preferable for this. I'd expect such analyzer look for all null literals and null checks (including `?.`, `??`, `??=`, etc) and offer to change the target type. Now, the said type could come from anywhere like a local declaration (if not var), parameter, field, property, class base, or a type argument. The fix-all could be a little more involved since we probably should infer nullability when you pass a value *to* a nullable parameter (need to guard for recursions). I think it'd make more sense to get this info from the compiler (#30110?)